{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/miNts3524/Bioinformatics/blob/main/Af_validation_dist.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "4N-ODEjRe28O"
      },
      "outputs": [],
      "source": [
        "%%time\n",
        "#@title setup **RFdiffusion** (~3min)\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "\n",
        "\n",
        "import os, time, signal\n",
        "import sys, random, string, re\n",
        "if not os.path.isdir(\"params\"):\n",
        "  os.system(\"apt-get install aria2\")\n",
        "  os.system(\"mkdir params\")\n",
        "  # send param download into background\n",
        "  os.system(\"(\\\n",
        "  aria2c -q -x 16 https://files.ipd.uw.edu/krypton/schedules.zip; \\\n",
        "  aria2c -q -x 16 https://storage.googleapis.com/alphafold/alphafold_params_2022-12-06.tar; \\\n",
        "  tar -xf alphafold_params_2022-12-06.tar -C params; \\\n",
        "  touch params/done.txt) &\")\n",
        "\n",
        "\"\"\"\n",
        "  os.system(\"(\\\n",
        "  aria2c -q -x 16 https://files.ipd.uw.edu/krypton/schedules.zip; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/6f5902ac237024bdd0c176cb93063dc4/Base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/e29311f6f1bf1af907f9ef9f44b8328b/Complex_base_ckpt.pt; \\\n",
        "  aria2c -q -x 16 http://files.ipd.uw.edu/pub/RFdiffusion/f572d396fae9206628714fb2ce00f72e/Complex_beta_ckpt.pt; \\\n",
        "  aria2c -q -x 16 https://storage.googleapis.com/alphafold/alphafold_params_2022-12-06.tar; \\\n",
        "  tar -xf alphafold_params_2022-12-06.tar -C params; \\\n",
        "  touch params/done.txt) &\")\n",
        "\"\"\"\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion\"):\n",
        "  print(\"installing RFdiffusion...\")\n",
        "  os.system(\"git clone https://github.com/sokrypton/RFdiffusion.git\")\n",
        "  os.system(\"pip install jedi omegaconf hydra-core icecream pyrsistent pynvml decorator\")\n",
        "  os.system(\"pip install git+https://github.com/NVIDIA/dllogger#egg=dllogger\")\n",
        "  # 17Mar2024: adding --no-dependencies to avoid installing nvidia-cuda-* dependencies\n",
        "  os.system(\"pip install --no-dependencies dgl==2.0.0 -f https://data.dgl.ai/wheels/cu121/repo.html\")\n",
        "  os.system(\"pip install --no-dependencies e3nn==0.3.3 opt_einsum_fx\")\n",
        "  os.system(\"cd RFdiffusion/env/SE3Transformer; pip install .\")\n",
        "  os.system(\"wget -qnc https://files.ipd.uw.edu/krypton/ananas\")\n",
        "  os.system(\"chmod +x ananas\")\n",
        "\n",
        "if not os.path.isdir(\"colabdesign\"):\n",
        "  print(\"installing ColabDesign...\")\n",
        "  os.system(\"pip -q install git+https://github.com/sokrypton/ColabDesign.git\")\n",
        "  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabdesign colabdesign\")\n",
        "\n",
        "if not os.path.isdir(\"RFdiffusion/models\"):\n",
        "  print(\"downloading RFdiffusion params...\")\n",
        "  os.system(\"mkdir RFdiffusion/models\")\n",
        "  models = [\"Base_ckpt.pt\",\"Complex_base_ckpt.pt\",\"Complex_beta_ckpt.pt\"]\n",
        "  for m in models:\n",
        "    while os.path.isfile(f\"{m}.aria2\"):\n",
        "      time.sleep(5)\n",
        "  os.system(f\"mv {' '.join(models)} RFdiffusion/models\")\n",
        "  os.system(\"unzip schedules.zip; rm schedules.zip\")\n",
        "\n",
        "if 'RFdiffusion' not in sys.path:\n",
        "  os.environ[\"DGLBACKEND\"] = \"pytorch\"\n",
        "  sys.path.append('RFdiffusion')\n",
        "\n",
        "from google.colab import files\n",
        "import json\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "from IPython.display import display, HTML\n",
        "import ipywidgets as widgets\n",
        "import py3Dmol\n",
        "\n",
        "from inference.utils import parse_pdb\n",
        "from colabdesign.rf.utils import get_ca\n",
        "from colabdesign.rf.utils import fix_contigs, fix_partial_contigs, fix_pdb, sym_it\n",
        "from colabdesign.shared.protein import pdb_to_string\n",
        "from colabdesign.shared.plot import plot_pseudo_3D\n",
        "\n",
        "def get_pdb(pdb_code=None):\n",
        "  if pdb_code is None or pdb_code == \"\":\n",
        "    upload_dict = files.upload()\n",
        "    pdb_string = upload_dict[list(upload_dict.keys())[0]]\n",
        "    with open(\"tmp.pdb\",\"wb\") as out: out.write(pdb_string)\n",
        "    return \"tmp.pdb\"\n",
        "  elif os.path.isfile(pdb_code):\n",
        "    return pdb_code\n",
        "  elif len(pdb_code) == 4:\n",
        "    if not os.path.isfile(f\"{pdb_code}.pdb1\"):\n",
        "      os.system(f\"wget -qnc https://files.rcsb.org/download/{pdb_code}.pdb1.gz\")\n",
        "      os.system(f\"gunzip {pdb_code}.pdb1.gz\")\n",
        "    return f\"{pdb_code}.pdb1\"\n",
        "  else:\n",
        "    os.system(f\"wget -qnc https://alphafold.ebi.ac.uk/files/AF-{pdb_code}-F1-model_v3.pdb\")\n",
        "    return f\"AF-{pdb_code}-F1-model_v3.pdb\"\n",
        "\n",
        "\n",
        "from string import ascii_uppercase, ascii_lowercase\n",
        "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
        "import os\n",
        "\n",
        "from Bio.PDB import PDBParser, Polypeptide\n",
        "\n",
        "def get_info(contig):\n",
        "  F = []\n",
        "  free_chain = False\n",
        "  fixed_chain = False\n",
        "  sub_contigs = [x.split(\"-\") for x in contig.split(\"/\")]\n",
        "  for n,(a,b) in enumerate(sub_contigs):\n",
        "    if a[0].isalpha():\n",
        "      L = int(b)-int(a[1:]) + 1\n",
        "      F += [1] * L\n",
        "      fixed_chain = True\n",
        "    else:\n",
        "      L = int(b)\n",
        "      F += [0] * L\n",
        "      free_chain = True\n",
        "  return F,[fixed_chain,free_chain]\n",
        "\n",
        "\n",
        "\n",
        "def three_to_one(aa):\n",
        "    \"\"\"3-letter 아미노산 코드를 1-letter 코드로 변환\"\"\"\n",
        "    aa_three_to_one = {\n",
        "        'ALA': 'A', 'ARG': 'R', 'ASN': 'N', 'ASP': 'D', 'CYS': 'C',\n",
        "        'GLN': 'Q', 'GLU': 'E', 'GLY': 'G', 'HIS': 'H', 'ILE': 'I',\n",
        "        'LEU': 'L', 'LYS': 'K', 'MET': 'M', 'PHE': 'F', 'PRO': 'P',\n",
        "        'SER': 'S', 'THR': 'T', 'TRP': 'W', 'TYR': 'Y', 'VAL': 'V',\n",
        "        'SEC': 'U', 'PYL': 'O', 'ASX': 'B', 'GLX': 'Z', 'XLE': 'J',\n",
        "        'XAA': 'X', 'UNK': 'X'\n",
        "    }\n",
        "    return aa_three_to_one.get(aa, 'X')\n",
        "\n",
        "class PDBInfo:\n",
        "    def __init__(self, pdb_file_path, target_chain_ids):\n",
        "        self.pdb_file_path = pdb_file_path\n",
        "        self.target_chain_ids = target_chain_ids\n",
        "        self.chain_info = self._parse_pdb(pdb_file_path)\n",
        "        self.prep_flags, self.protocol = self._generate_prep_flags()\n",
        "\n",
        "    def _parse_pdb(self, pdb_file_path):\n",
        "        \"\"\"PDB 파일을 파싱하고 체인 정보를 추출\"\"\"\n",
        "        parser = PDBParser(QUIET=True)\n",
        "        structure = parser.get_structure(\"protein\", pdb_file_path)\n",
        "\n",
        "        chain_info = {}\n",
        "\n",
        "        for chain in structure[0]:\n",
        "            chain_id = chain.id\n",
        "            full_sequence = []\n",
        "            sequence_parts = []\n",
        "            current_segment = []\n",
        "            previous_residue_id = None\n",
        "            for residue in chain:\n",
        "                if Polypeptide.is_aa(residue, standard=True):\n",
        "                    current_residue_id = residue.id[1]\n",
        "                    full_sequence.append(three_to_one(residue.resname))\n",
        "\n",
        "                    if previous_residue_id is None or current_residue_id == previous_residue_id + 1:\n",
        "                        current_segment.append(current_residue_id)\n",
        "                    else:\n",
        "                        if current_segment:\n",
        "                            sequence_parts.append((current_segment[0], current_segment[-1]))\n",
        "                        current_segment = [current_residue_id]\n",
        "\n",
        "                    previous_residue_id = current_residue_id\n",
        "\n",
        "            if current_segment:\n",
        "                sequence_parts.append((current_segment[0], current_segment[-1]))\n",
        "\n",
        "            summary = \"/\".join([f\"{start}-{end}\" for start, end in sequence_parts])\n",
        "            chain_info[chain_id] = {\n",
        "                \"chain_id\": chain_id,\n",
        "                \"chain_seq\": \"\".join(full_sequence),\n",
        "                \"summary\": summary,\n",
        "                \"length\": len(full_sequence)\n",
        "            }\n",
        "\n",
        "        return chain_info\n",
        "\n",
        "    def _generate_prep_flags(self):\n",
        "        \"\"\"PDB 파일에 대한 준비 플래그를 생성\"\"\"\n",
        "\n",
        "        contig_str = \":\".join(self.get_contigs())\n",
        "        contigs = []\n",
        "        for contig_str in contig_str.replace(\" \", \":\").replace(\",\", \":\").split(\":\"):\n",
        "            if len(contig_str) > 0:\n",
        "                contig = []\n",
        "                for x in contig_str.split(\"/\"):\n",
        "                    if x != \"0\": contig.append(x)\n",
        "                contigs.append(\"/\".join(contig))\n",
        "\n",
        "        chains = alphabet_list[:len(contigs)]\n",
        "        info = [get_info(x) for x in contigs]\n",
        "        fixed_pos = []\n",
        "        fixed_chains = []\n",
        "        free_chains = []\n",
        "        both_chains = []\n",
        "        for pos, (fixed_chain, free_chain) in info:\n",
        "            fixed_pos += pos\n",
        "            fixed_chains += [fixed_chain and not free_chain]\n",
        "            free_chains += [free_chain and not fixed_chain]\n",
        "            both_chains += [fixed_chain and free_chain]\n",
        "\n",
        "        if sum(both_chains) == 0 and sum(fixed_chains) > 0 and sum(free_chains) > 0 :\n",
        "            protocol = \"binder\"\n",
        "            target_chains = []\n",
        "            binder_chains = []\n",
        "            for n,x in enumerate(fixed_chains):\n",
        "                if x: target_chains.append(chains[n])\n",
        "                else: binder_chains.append(chains[n])\n",
        "            prep_flags = {\"target_chain\": \",\".join(target_chains),\n",
        "                          \"binder_chain\": \",\".join(binder_chains),\n",
        "                          \"rm_aa\": 'None'}\n",
        "        return prep_flags, protocol\n",
        "\n",
        "\n",
        "    def target_info(self):\n",
        "        \"\"\"타겟 체인 정보를 반환\"\"\"\n",
        "        return {\n",
        "            \"pdb_path\": self.pdb_file_path,\n",
        "            \"chains\": {cid: self.chain_info[cid] for cid in self.target_chain_ids if cid in self.chain_info}\n",
        "        }\n",
        "\n",
        "    def chain(self, chain_id):\n",
        "        \"\"\"특정 체인의 정보를 반환\"\"\"\n",
        "        return self.chain_info.get(chain_id, None)\n",
        "\n",
        "    def all_chains(self):\n",
        "        \"\"\"모든 체인의 정보를 반환\"\"\"\n",
        "        return self.chain_info\n",
        "\n",
        "    def all_sequences(self):\n",
        "        \"\"\"모든 체인의 서열을 '/'로 구분하여 하나의 문자열로 반환\"\"\"\n",
        "        return \"/\".join([self.chain_info[chain_id][\"chain_seq\"] for chain_id in self.chain_info])\n",
        "\n",
        "    def get_contigs(self):\n",
        "        \"\"\"target 체인에 대한 contigs를 반환\"\"\"\n",
        "        contigs = []\n",
        "\n",
        "        # binder (A 체인)의 길이를 추가\n",
        "        binder_info = self.chain_info.get('A')\n",
        "        if binder_info:\n",
        "            binder_length = binder_info[\"length\"]\n",
        "            contigs.append(f\"{binder_length}-{binder_length}\")\n",
        "\n",
        "        for chain_id in self.target_chain_ids:\n",
        "            chain_info = self.chain_info.get(chain_id)\n",
        "            if chain_info:\n",
        "                summary = chain_info[\"summary\"]\n",
        "                parts = summary.split(\"/\")\n",
        "                contig_parts = [f\"{chain_id}{part}\" for part in parts]\n",
        "                contigs.append(\"/\".join(contig_parts))\n",
        "        return contigs\n",
        "    def get_prep_flags(self):\n",
        "        return self.prep_flags\n",
        "\n",
        "\n",
        "class PDBManager:\n",
        "    def __init__(self, input_dir, target_chain_ids):\n",
        "        self.input_dir = input_dir\n",
        "        self.target_chain_ids = target_chain_ids\n",
        "        self.pdb_info_list = self._load_pdb_files()\n",
        "\n",
        "    def _load_pdb_files(self):\n",
        "        pdb_info_list = []\n",
        "        pdb_files = [os.path.join(self.input_dir, f) for f in os.listdir(self.input_dir) if f.endswith(\".pdb\")]\n",
        "        print(f\"Found PDB files: {pdb_files}\")  # 디버깅을 위해 추가\n",
        "        for pdb_file in pdb_files:\n",
        "            print(f\"Processing PDB file: {pdb_file}\")  # 디버깅을 위해 추가\n",
        "            pdb_info = PDBInfo(pdb_file, self.target_chain_ids)\n",
        "            pdb_info_list.append(pdb_info)\n",
        "        return pdb_info_list\n",
        "\n",
        "    def get_pdb_info(self, pdb_file_path):\n",
        "        for pdb_info in self.pdb_info_list:\n",
        "            if pdb_info.pdb_file_path == pdb_file_path:\n",
        "                return pdb_info\n",
        "        return None\n",
        "\n",
        "    def get_all_pdbs(self):\n",
        "        return self.pdb_info_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xXb5IC3zftMF",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Install AF\n",
        "import os,sys\n",
        "from colabdesign.af import mk_af_model\n",
        "from colabdesign.shared.protein import pdb_to_string\n",
        "from colabdesign.shared.parse_args import parse_args\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from google.colab import data_table\n",
        "data_table.enable_dataframe_formatter()\n",
        "\n",
        "def designable(target_chain_list, arg_rm_aa, arg_copies, arg_initial_guess, arg_use_multimer, arg_num_seqs, arg_num_recycles, input_dir):\n",
        "    pdb_manager = PDBManager(input_dir, target_chain_list)\n",
        "\n",
        "    flags = {\"initial_guess\": arg_initial_guess,\n",
        "                \"best_metric\": \"rmsd\",\n",
        "                \"use_multimer\": arg_use_multimer,\n",
        "                \"model_names\": [\"model_1_multimer_v3\" if arg_use_multimer else \"model_1_ptm\"]}\n",
        "\n",
        "    ###################################need to fix#####################################################\n",
        "    protocol = 'binder'\n",
        "    af_model = mk_af_model(protocol, **flags)\n",
        "    #####################Running AF#####################\n",
        "    if protocol == \"binder\":\n",
        "        af_terms = [\"plddt\",\"i_ptm\",\"i_pae\",\"rmsd\"]\n",
        "    elif arg_copies > 1:\n",
        "        af_terms = [\"plddt\",\"ptm\",\"i_ptm\",\"pae\",\"i_pae\",\"rmsd\"]\n",
        "    else:\n",
        "        af_terms = [\"plddt\",\"ptm\",\"pae\",\"rmsd\"]\n",
        "\n",
        "\n",
        "    labels = [\"design\",\"n\",\"score\"] + af_terms + [\"seq\"]\n",
        "\n",
        "    print(\"running AlphaFold...\")\n",
        "\n",
        "\n",
        "    output_dir = f\"{input_dir}/Af_validation\"\n",
        "    os.system(f\"mkdir -p {output_dir}/all_pdb\")\n",
        "\n",
        "    #os.system(f\"mkdir -p {output_dir}\")\n",
        "\n",
        "    outs = []\n",
        "    pdbs = pdb_manager.get_all_pdbs()\n",
        "    filtered = []\n",
        "    data = []\n",
        "    binder = {\"data_binder\" : [], \"binder_path\" : [], \"binder_input\" : []}\n",
        "    with open(f\"{output_dir}/design.fasta\",\"w\") as fasta:\n",
        "        count = 1\n",
        "        for m, pdb in enumerate(pdbs):\n",
        "            out = {\"design\": [], \"n\": [], \"score\": [], \"seq\": []}\n",
        "            pdb_path = pdb.pdb_file_path\n",
        "            total_sequence = pdb.all_sequences()\n",
        "            prep_flags = pdb.get_prep_flags()\n",
        "            prep_flags[\"rm_aa\"] = arg_rm_aa\n",
        "\n",
        "            ###########################get_seqs.\n",
        "            if protocol == 'binder':\n",
        "\n",
        "                binder_sequence = pdb.chain_info['A']['chain_seq']\n",
        "            else:\n",
        "                error_message = \"Not binder protocol\"\n",
        "                raise ValueError(error_message)\n",
        "\n",
        "            for k in af_terms:\n",
        "              out[k] = []\n",
        "            #for n in range(arg_num_seqs):\n",
        "            af_model.prep_inputs(pdb.pdb_file_path, **prep_flags)\n",
        "            out[\"design\"] = m\n",
        "            out[\"score\"] = 0\n",
        "            out[\"seq\"] = total_sequence\n",
        "\n",
        "            #sub_seq = out[\"seq\"].replace(\"/\",\"\")[af_model._len:]\n",
        "\n",
        "            sub_seq = binder_sequence\n",
        "            af_model.predict(seq=sub_seq, num_recycles=arg_num_recycles, verbose=False)\n",
        "            for t in af_terms:\n",
        "                out[t] = af_model.aux[\"log\"][t]\n",
        "            for t in af_terms:\n",
        "                out[t] = af_model.aux[\"log\"][t]\n",
        "            if \"i_pae\" in out:\n",
        "                out[\"i_pae\"] = out[\"i_pae\"] * 31\n",
        "            if \"pae\" in out:\n",
        "                out[\"pae\"] = out[\"pae\"] * 31\n",
        "            rmsd = out[\"rmsd\"]\n",
        "\n",
        "\n",
        "            # Get binder\n",
        "            if out[\"i_pae\"] < 10 and out[\"rmsd\"] < 1 and out[\"plddt\"] > 0.8:\n",
        "                print(\"Congraturations!! Binder was Found!!! ⤵⤵⤵\")\n",
        "                af_model.save_current_pdb(f\"{output_dir}/binder_{m}.pdb\")\n",
        "                binder[\"data_binder\"].append([out[k] for k in labels])\n",
        "                binder[\"binder_path\"].append(f\"{output_dir}/binder_{m}.pdb\")\n",
        "                binder[\"binder_input\"].append(f\"{pdb.pdb_file_path}\")\n",
        "\n",
        "            af_model.save_current_pdb(f\"{output_dir}/all_pdb/design_{m}.pdb\")\n",
        "            af_model._save_results(save_best=True, verbose=False)\n",
        "            af_model._k += 1\n",
        "            score_line = [f'design : {m}']\n",
        "            for t in af_terms:\n",
        "                score_line.append(f'{t} : {out[t]:.3f}')\n",
        "            line = f'>{\"|\".join(score_line)}\\n{out[\"seq\"]}'\n",
        "            fasta.write(line+\"\\n\")\n",
        "            outs.append(out)\n",
        "            #print(\" \".join(score_line)+\" \"+out[\"seq\"])\n",
        "            data.append([out[k] for k in labels])\n",
        "            print(f\"Design : {m}, plddt : {out['plddt']:.3f}, rmsd : {out['rmsd']:.3f}, i_pae : {out['i_pae']:.3f}\")\n",
        "            print(f\"Sequence : {out['seq']}\")\n",
        "\n",
        "\n",
        "    df = pd.DataFrame(data, columns = labels)\n",
        "    df.to_csv(f'{output_dir}/AF_valid.csv')\n",
        "    df_binder = pd.DataFrame(binder['data_binder'], columns = labels)\n",
        "    df_binder.to_csv(f'{output_dir}/AF_binder.csv')\n",
        "    data_table.DataTable(df.sort_values(\"rmsd\").round(3))\n",
        "    #print(df.sort_values(\"rmsd\").round(3))\n",
        "\n",
        "\n",
        "    return binder, df , labels\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "n_i0C6aRoGuS"
      },
      "outputs": [],
      "source": [
        "%%time\n",
        "#@title run **AlphaFold** to validate\n",
        "\n",
        "\n",
        "#drive/MyDrive/pdb_files/hotspot_mpnn_step2\n",
        "\n",
        "#@markdown INPUT directory -- drive/MyDrive\n",
        "input_dir = \"drive/MyDrive/PLA/test\" #@param {type:\"string\"}\n",
        "#output_dir = 'drive/MyDrive/pdb_files/hotpot' #@param {type:\"string\"}\n",
        "\n",
        "#@markdown ProteinMPNN Settings\n",
        "\n",
        "Target_chain = 'B' #@param {type:\"string\"}\n",
        "target_chain_list = Target_chain.split(\",\")\n",
        "print(target_chain_list)\n",
        "\n",
        "num_seqs = 1 #@param [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\", \"64\"] {type:\"raw\"}\n",
        "#mpnn_sampling_temp = 0.1 #@param [\"0.0001\", \"0.1\", \"0.15\", \"0.2\", \"0.25\", \"0.3\", \"0.5\", \"1.0\"] {type:\"raw\"}\n",
        "rm_aa = \"C\" #@param {type:\"string\"}\n",
        "use_solubleMPNN = False #@param {type:\"boolean\"}\n",
        "#@markdown - `Target_chain` - list chains. (ex. A,B) / caution **No space**\n",
        "#@markdown - `mpnn_sampling_temp` - control diversity of sampled sequences. (higher = more diverse).\n",
        "#@markdown - `rm_aa='C'` - do not use [C]ysteines.\n",
        "#@markdown - `use_solubleMPNN` - use weights trained only on soluble proteins. See [preprint](https://www.biorxiv.org/content/10.1101/2023.05.09.540044v2).\n",
        "#@markdown\n",
        "#@markdown AlphaFold Settings\n",
        "initial_guess = True #@param {type:\"boolean\"}\n",
        "#@markdown - soft initialization with desired coordinates, see [paper](https://www.nature.com/articles/s41467-023-38328-5).\n",
        "num_recycles = 3 #@param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\"] {type:\"raw\"}\n",
        "#@markdown - for **binder** design, we recommend `initial_guess=True num_recycles=3`\n",
        "use_multimer = True #@param {type:\"boolean\"}\n",
        "#@markdown - `use_multimer` - use AlphaFold Multimer v3 params for prediction.\n",
        "\n",
        "\n",
        "#Getting PDB\n",
        "# for pdb_info in pdb_manager.get_all_pdbs():\n",
        "#     print(f\"PDB Path: {pdb_info.pdb_file_path}\")\n",
        "#     print(\"Target Chains:\")\n",
        "#     for chain_id, details in pdb_info.target_info()[\"chains\"].items():\n",
        "#         print(f\"  Chain ID: {details['chain_id']}\")\n",
        "#         print(f\"  Chain Sequence: {details['chain_seq']}\")\n",
        "#         print(f\"  Summary: {details['summary']}\")\n",
        "#     print(\"All Chains:\")\n",
        "#     for chain_id, details in pdb_info.all_chains().items():\n",
        "#         print(f\"  Chain ID: {details['chain_id']}\")\n",
        "#         print(f\"  Chain Sequence: {details['chain_seq']}\")\n",
        "#         print(f\"  Summary: {details['summary']}\")\n",
        "#     print(f\"All Sequences: {pdb_info.all_sequences()}\")\n",
        "#     print(\"Contigs: \", pdb_info.get_contigs())\n",
        "#     print(\"Prep Flags: \", pdb_info.get_prep_flags())\n",
        "#     print()\n",
        "\n",
        "\n",
        "# output_folder = os.path.join(input_dir, \"AF_all_pdb\")\n",
        "\n",
        "# #pdb_list = [os.path.join(input_dir,x) for x in os.listdir(input_dir) if x.endswith(\".pdb\")]\n",
        "# target_chain_list = Target_chain.split(\",\")\n",
        "# #print(target_chain_list)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "copies = 1 #binder\n",
        "\n",
        "binder, df, labels = designable(target_chain_list,\n",
        "                    rm_aa,\n",
        "                    copies,\n",
        "                    initial_guess,\n",
        "                    use_multimer,\n",
        "                    num_seqs,\n",
        "                    num_recycles,\n",
        "                    input_dir)\n",
        "\n",
        "df_total['binder_length'] = df['seq'].apply(lambda x: len(x.split(\"/\")[0]))\n",
        "plt.figure(figsize=(10, 6))\n",
        "scatter = plt.scatter(df_total['binder_length'],\n",
        "                      df_total['rmsd'],\n",
        "                      c=df_total['plddt'],\n",
        "                      cmap='viridis',\n",
        "                      alpha=0.7)\n",
        "\n",
        "plt.colorbar(scatter, label='plddt')\n",
        "plt.xlabel('Sequence Length')\n",
        "plt.ylabel('RMSD')\n",
        "plt.title('Scatter Plot of Sequence Length vs. RMSD with pLDDT Coloring')\n",
        "plt.savefig(f'{input_dir}/Af_validation/scatter_plot.png')\n",
        "\n",
        "\n",
        "# result_df = pd.DataFrame(binder['data_binder'], columns=labels)\n",
        "# print(result_df)\n",
        "# data_table.DataTable(result_df.sort_values(\"rmsd\").round(3))\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Total result summary : table\n",
        "\n",
        "df_total =pd.DataFrame(df, columns=labels)\n",
        "df_total = df_total.drop(['n', 'design','score'], axis = 'columns')\n",
        "data_table.DataTable(df_total.sort_values(\"rmsd\").round(3))"
      ],
      "metadata": {
        "cellView": "form",
        "id": "2sslfJl_zudv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Total result summary : Plot\n",
        "\n",
        "df_total['binder_length'] =df['seq'].apply(lambda x: len(x.split(\"/\")[0]))\n",
        "plt.figure(figsize=(10, 6))\n",
        "scatter = plt.scatter(df_total['binder_length'],\n",
        "                      df_total['rmsd'],\n",
        "                      c=df_total['plddt'],\n",
        "                      cmap='viridis',\n",
        "                      alpha=0.7)\n",
        "\n",
        "plt.colorbar(scatter, label='plddt')\n",
        "plt.xlabel('Sequence Length')\n",
        "plt.ylabel('RMSD')\n",
        "plt.title('Scatter Plot of Sequence Length vs. RMSD with pLDDT Coloring')\n",
        "#plt.savefig(f'{input_dir}/Af_validation/scatter_plot.png')\n",
        "plt.show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "gazLy5qJ3nEK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "HUQ9HCtdW2mO",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Display Binder result\n",
        "import py3Dmol\n",
        "\n",
        "def plot_pdb(binder, binder_input):\n",
        "  with open(binder,\"r\") as f:\n",
        "    # REMARK 001 design {m} N {n} RMSD {rmsd}\n",
        "    #info = f.readline().strip('\\n').split()\n",
        "    hbondCutoff = 4.0\n",
        "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
        "    pdb_str = open(binder_input,'r').read()\n",
        "    view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "    pdb_str = open(binder,'r').read()\n",
        "    view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "\n",
        "    view.setStyle({\"model\":0},{'cartoon':{}}) #: {'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
        "    view.setStyle({\"model\":1},{'cartoon':{'colorscheme': {'prop':'b','gradient': 'roygb','min':0,'max':100}}})\n",
        "    view.zoomTo()\n",
        "    view.show()\n",
        "\n",
        "                # binder[\"data_binder\"].append([out[k] for k in labels])\n",
        "                # binder[\"binder_path\"].append(f\"{output_dir}/Af_validation/binder_{m}.pdb\")\n",
        "                # binder[\"binder_input\"].append(f\"{pdb.pdb_file_path}\")\n",
        "\n",
        "\n",
        "# if len(binder[\"binder_path\"]) < 1:\n",
        "#   print(\"Binder was not found!\")\n",
        "\n",
        "# if len(binder) > 1:\n",
        "#   print(f\"{len(binder['binder_path'])} binder was found!\")\n",
        "#   output = widgets.Output()\n",
        "#   def on_change(change):\n",
        "#     if change['name'] == 'value':\n",
        "#       with output:\n",
        "#         output.clear_output(wait=True)\n",
        "#         plot_pdb(change['new'])\n",
        "#   dropdown = widgets.Dropdown(\n",
        "#       options=[(f'{k}',k) for k in range(len(binder['binder_path']))],\n",
        "#       value=0, description='design:',\n",
        "#   )\n",
        "#   dropdown.observe(on_change)\n",
        "#   display(widgets.VBox([dropdown, output]))\n",
        "#   with output:\n",
        "#     plot_pdb(dropdown.value)\n",
        "# else:\n",
        "#   plot_pdb()\n",
        "if len(binder[\"binder_path\"]) < 1:\n",
        "    print(\"Binder was not found!\")\n",
        "else:\n",
        "    if len(binder[\"binder_path\"]) > 1:\n",
        "        print(f\"{len(binder['binder_path'])} binders were found!\")\n",
        "        output = widgets.Output()\n",
        "\n",
        "        def on_change(change):\n",
        "            if change['name'] == 'value':\n",
        "                with output:\n",
        "                    output.clear_output(wait=True)\n",
        "                    plot_pdb(binder['binder_path'][change['new']], binder['binder_input'][change['new']])\n",
        "\n",
        "        dropdown = widgets.Dropdown(\n",
        "            options=[(f'{k}', k) for k in range(len(binder['binder_path']))],\n",
        "            value=0, description='design:'\n",
        "        )\n",
        "        dropdown.observe(on_change, names='value')\n",
        "        display(widgets.VBox([dropdown, output]))\n",
        "\n",
        "        with output:\n",
        "            plot_pdb(binder['binder_path'][0], binder['binder_input'][0])\n",
        "    else:\n",
        "        plot_pdb(binder['binder_path'][0], binder['binder_input'][0])\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMeNhEI6BBP9dtDrghguzPO",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}